examples/structured_data/collaborative_filtering_movielens.py

```diff
@@ -0,0 +1,226 @@
+userencoded2user = {i: x for i, x in enumerate(user_ids)}
+movie_ids = df["movieId"].unique().tolist()
+movie2movie_encoded = {x: i for i, x in enumerate(movie_ids)}
+movie_encoded2movie = {i: x for i, x in enumerate(movie_ids)}
+df["user"] = df["userId"].map(user2user_encoded)
+df["movie"] = df["movieId"].map(movie2movie_encoded)
+
+num_users = len(user2user_encoded)
+num_movies = len(movie_encoded2movie)
+df["rating"] = df["rating"].values.astype(np.float32)
+min_rating = min(df["rating"])
+max_rating = max(df["rating"])
+print(
+    "Number of users: {}, Number of Movies: {}, Min rating: {}, Max rating: {}".format(
+        num_users, num_movies, min_rating, max_rating
+    )
+)
+
+"""
+## Prepare training and validation data
+"""
+df = df.sample(frac=1, random_state=42)
+x = df[["user", "movie"]].values
+# Normalize the targets between 0 and 1. Makes it easy to train.
+y = df["rating"].apply(lambda x: (x - min_rating) / (max_rating - min_rating)).values
+# Assuming training on 90% of the data and validating on 10%.
+train_indices = int(0.9 * df.shape[0])
+x_train, x_test, y_train, y_test = (
+    x[:train_indices],
+    x[train_indices:],
+    y[:train_indices],
+    y[train_indices:],
+)
+
+"""
+## Create the model
+
+Let's fix some parameters in the architecture. We will use user embeddings and movie
+embeddings, let's keep the dimensions as 50 for both.
+The model computes a score using dot product of user and movie embeddings.
+Thereafter, it uses a sigmoid activation on the sum over all the scores (including biases).
+Finally, the score is converted into the rating ranges based on this dataset.
+"""
+EMBEDDING_SIZE = 50
+
+
+class RecommenderNet(keras.Model):
+    def __init__(
+        self, num_users, num_movies, embedding_size, min_rating, max_rating, **kwargs
+    ):
+        super(RecommenderNet, self).__init__(**kwargs)
+        self.num_users = num_users
+        self.num_movies = num_movies
+        self.embedding_size = embedding_size
+        self.min_rating = min_rating
+        self.max_rating = max_rating
+        self.user_embedding = layers.Embedding(
+            num_users,
+            embedding_size,
+            embeddings_initializer="he_normal",
+            embeddings_regularizer=keras.regularizers.l2(1e-6),
+        )
+        self.user_bias = layers.Embedding(num_users, 1)
+        self.movie_embedding = layers.Embedding(
+            num_movies,
+            embedding_size,
+            embeddings_initializer="he_normal",
+            embeddings_regularizer=keras.regularizers.l2(1e-6),
+        )
+        self.movie_bias = layers.Embedding(num_movies, 1)
+
+    def call(self, inputs):
+        user_vector = self.user_embedding(inputs[:, 0])
+        user_bias = self.user_bias(inputs[:, 0])
+        movie_vector = self.movie_embedding(inputs[:, 1])
+        movie_bias = self.movie_bias(inputs[:, 1])
+        dot_user_movie = layers.Dot(axes=1)([user_vector, movie_vector])
```

Comment:
Why not use `tf.tensordot` here?

If you want to use `Dot`, you should instantiate the layer in `__init__` and call it here (otherwise you are instantiating a new instance every time you call the model)

Reply:
Actually tf.tensordot was taking a very long time because my conda tensorflow installation had some issues. I fixed it and will use it. Will submit the updated code soon. Thanks very much!